src/meta/src/stream/stream_graph.rs

```diff
@@ -624,37 +498,78 @@ impl StreamGraphBuilder {
 }
 
 /// The mutable state when building actor graph.
-#[derive(Default)]
 struct BuildActorGraphState {
-    /// stream graph builder, to build streaming DAG.
+    /// The stream graph builder, to build streaming DAG.
     stream_graph_builder: StreamGraphBuilder,
-    /// when converting fragment graph to actor graph, we need to know which actors belong to a
+
+    /// When converting fragment graph to actor graph, we need to know which actors belong to a
     /// fragment.
-    fragment_actors: HashMap<GlobalFragmentId, Vec<LocalActorId>>,
-    /// local actor id
-    next_local_actor_id: u32,
+    fragment_actors: HashMap<GlobalFragmentId, Vec<GlobalActorId>>,
+
+    /// The next local actor id to use.
+    next_local_id: u32,
+
+    /// The global actor id generator.
+    actor_id_gen: GlobalActorIdGen,
 }
 
 impl BuildActorGraphState {
-    fn gen_actor_ids(&mut self, parallel_degree: u32) -> Range<u32> {
-        let start_actor_id = self.next_local_actor_id;
-        self.next_local_actor_id += parallel_degree;
-        start_actor_id..start_actor_id + parallel_degree
+    /// Create an empty state with the given id generator.
+    fn new(actor_id_gen: GlobalActorIdGen) -> Self {
+        Self {
+            stream_graph_builder: Default::default(),
+            fragment_actors: Default::default(),
+            next_local_id: 0,
+            actor_id_gen,
+        }
+    }
+
+    /// Get the next global actor id.
+    fn next_actor_id(&mut self) -> GlobalActorId {
+        let local_id = self.next_local_id;
+        self.next_local_id += 1;
+
+        self.actor_id_gen.to_global_id(local_id)
+    }
+
+    /// Finish the build and return the inner stream graph builder.
+    fn finish(self) -> StreamGraphBuilder {
+        assert_eq!(self.actor_id_gen.len, self.next_local_id);
+        self.stream_graph_builder
     }
 }
 
 /// [`ActorGraphBuilder`] generates the proto for interconnected actors for a streaming pipeline.
 pub struct ActorGraphBuilder {
-    /// Default parallelism.
-    default_parallelism: u32,
+    /// The pre-scheduled distribution for each fragment.
+    // TODO: this is fake now and only the parallelism is used, we should also use the physical
+    // distribution itself after scheduler refactoring.
+    distributions: HashMap<GlobalFragmentId, Distribution>,
 
+    /// The stream fragment graph.
     fragment_graph: StreamFragmentGraph,
 }
 
 impl ActorGraphBuilder {
-    pub fn new(fragment_graph: StreamFragmentGraph, default_parallelism: u32) -> Self {
+    /// Create a new actor graph builder with the given "complete" graph.
+    pub fn new(complete_graph: CompleteStreamFragmentGraph, default_parallelism: u32) -> Self {
+        // TODO: use the real parallel units to generate real distribution.
+        let fake_parallel_units = (0..default_parallelism).map(|id| ParallelUnit {
```

Comment:
As we only care about the parallelism of the scheduler result and the old scheduler is still working, we use fake parallel units here.